Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,640 @@
+        }
+        index--;
+    }
+
+    return NSOrderedSame;
+}
+
+/**
+ @Status Caveat
+ @Notes roundingMode is not supported.
+*/
+NSCalculationError NSDecimalNormalize(NSDecimal* number1, NSDecimal* number2, NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG,
+                      (number1 != NULL) && (number2 != NULL) && !NSDecimalIsNotANumber(number1) && !NSDecimalIsNotANumber(number2));
+
+    if (number1->_exponent == number2->_exponent) {
+        return NSCalculationNoError;
+    }
+
+    NSDecimal* decimalToNormalize;
+    NSDecimal* refDecimal;
+
+    if (number1->_exponent > number2->_exponent) {
+        decimalToNormalize = number1;
+        refDecimal = number2;
+    } else {
+        decimalToNormalize = number2;
+        refDecimal = number1;
+    }
+
+    int exponent = decimalToNormalize->_exponent - refDecimal->_exponent;
+
+    NSDecimalInternal internalResult = {};
+
+    NSDecimalCopy(&internalResult.decimal, decimalToNormalize);
+    internalResult._mantissa = internalResult.decimal._mantissa;
+    internalResult._remainder = 0;
+    internalResult._extendedLen = NSDecimalMaxSize;
+
+    NSCalculationError error = NSCalculationNoError;
+    // TODO: this can be better done, by creating a mantissa out of the 10^exp and doing mantissa by mantissa multiplication.
+    // This should be changed when NSDecimalMultiplication is implemented.
+    while (exponent > 0) {
+        error = _mantissaMultiplication(&internalResult, &internalResult, 10);
+        if (error != NSCalculationNoError) {
+            return error;
+        }
+        exponent--;
+    }
+
+    internalResult.decimal._exponent = refDecimal->_exponent;
+    _InternalDecimalCopy(decimalToNormalize, &internalResult);
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCompact(NSDecimal* number) {
+    if ((number == NULL) || NSDecimalIsNotANumber(number) || IS_ZERO(number)) {
+        return;
+    }
+
+    int exponent = number->_exponent;
+    NSDecimalInternal internalResult = {};
+
+    // reduce by 10 as much as we can
+    NSDecimalCopy(&internalResult.decimal, number);
+    internalResult._mantissa = internalResult.decimal._mantissa;
+    internalResult._remainder = 0;
+    internalResult._extendedLen = number->_length;
+
+    do {
+        _mantissaDivision(&internalResult, &internalResult, 10);
+        exponent++;
+    } while (!internalResult._remainder);
+
+    // put back the leftover reminder.
+    _mantissaMultiplication(&internalResult, &internalResult, 10);
+    exponent--;
+    _mantissaAddition(&internalResult, &internalResult, internalResult._remainder);
+
+    while (exponent > _MAX_EXP) {
+        // refit
+        _mantissaMultiplication(&internalResult, &internalResult, 10);
+        exponent--;
+    }
+
+    // update the exp
+    internalResult.decimal._exponent = exponent;
+    _InternalDecimalCopy(number, &internalResult);
+    number->_isCompact = 1;
```

Comment:
usually only time this could be false is if the user strictly passes in a custom decimal struct . The reason i'm not checking at the top for that exact case.
Just try to compact as much as we can (if it's already compacted , great

---
In reply to: [81226595](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r81226595) [](ancestors = 81226595)